### PR TITLE
Issue/85: JSON-RPC specs section closes #85

### LINKED ISSUE #85: JSON-RPC specs section
**Goal**: to provide a general overview of what is the JSON-RPC spec, how it works and add further reading links

### PR DESCRIPTION
**Pull request by rdr-x**
JSON-RPC specs section added. Included some method lists and a couple of tooling examples of how to use them through curl and axios (js/ts)

### CODE REVIEW

File: docs/wiki/EL/JSON-RPC.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
+# JSON-RPC
+
+---
+
+The JSON-RPC specification is a remote procedure call protocol encoded in JSON based on [OpenRPC](https://open-rpc.org/getting-started). It allows calling functions on a remote server, and for the return of the results.
+It is part of the Execution API specification which provides a set of methods to interact with the Ethereum blockchain.
+It is better known to be the way of how the clients interact with the network, even how the consensus layer (CL) and the execution layer (EL) interact.
```
Comment (taxmeifyoucan) [inline docs/wiki/EL/JSON-RPC.md:7]: > how the clients interact with the network

This can be misleading, I would rephrase to something like 'users interact with the network using a client'..

File: docs/wiki/EL/JSON-RPC.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+This section provides a description of the JSON-RPC methods.
+
+## JSON-RPC Methods
+
+There are a lot of JSON-RPC methods, but for summarizing purposes here are divided into the following prefix categories: *debug*, *engine* and *eth*, and all of them share a common structure:
```
Comment (taxmeifyoucan) [inline docs/wiki/EL/JSON-RPC.md:12]: I would highlight here there although there are more namespaces used by clients (admin, net, txpool, ots, etc), only these are specified and must behave the same across all implementations.

File: docs/wiki/EL/JSON-RPC.md (hunk 1 above)
Comment (taxmeifyoucan) [reply docs/wiki/EL/JSON-RPC.md:7]: Also when mentioning it's used for EL<>CL communication, I would explicitly said that engine API uses it as well but it's another authenticated interface not to be confused with user facing

File: docs/wiki/EL/JSON-RPC.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+| eth_getBalance           |  address, block number | returns the balance of the account of the given address          |
+| eth_getBlockByHash       |  block hash, full txs  | returns information about a block by hash                        |
+| eth_getBlockByNumber     |  block number, full txs | returns information about a block by block number                |
+| eth_getBlockTransactionCountByHash | block hash | returns the number of transactions in a block from a block matching the given block hash |
+| eth_getBlockTransactionCountByNumber | block number | returns the number of transactions in a block from a block matching the given block number |
+| eth_getCode              |  address, block number | returns code at a given address in the blockchain                 |
+| eth_getLogs              |  filter object         | returns an array of all logs matching a given filter object       |
+| eth_getStorageAt         |  address, position, block number | returns the value from a storage position at a given address |
+
+## Tooling
```
Comment (taxmeifyoucan) [inline docs/wiki/EL/JSON-RPC.md:74]: Before diving into tooling, I would explain that client can support different protocols for communication with the client - http, wss or ipc.

File: docs/wiki/EL/JSON-RPC.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    'Content-Type': 'application/json',
+  },
+});
+```
+As you may notice, the JSON-RPC methods are wrapped in a POST request, and the parameters are passed in the body of the request.
+This is a different way to exchange data between the client and the server using the OSI's application layer: the HTTP protocol.
+
+Usually, all the web3 libraries wrap the JSON-RPC methods providing a more friendly way to interact with the execution layer. Please, look forward in your preferred programming language as the syntax may vary.
+
+### Further Reading
```
Comment (taxmeifyoucan) [inline docs/wiki/EL/JSON-RPC.md:109]: great documentation of all calls which can be also an inspiration for you is at ethereumorg https://ethereum.org/en/developers/docs/apis/json-rpc

File: docs/wiki/EL/JSON-RPC.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+Where:
+- `id`: A unique identifier for the request.
+- `jsonrpc`: The version of the JSON-RPC protocol.
+- `method`: The method to be called.
+- `params`: The parameters for the method. It can be an empty array if the method does not require any parameters. Other ones may have default values if not provided.
+
+Here is a few list of the JSON-RPC methods:
+
+### Debug
```
Comment (taxmeifyoucan) [inline docs/wiki/EL/JSON-RPC.md:31]: I would give an overview of what is the namespace used for first

File: docs/wiki/EL/JSON-RPC.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+| **Method**               |      **Params**       | **Description**                                                 |
+|--------------------------|:---------------------:|-----------------------------------------------------------------|
+| debug_getBadBlocks       |  no mandatory params  | returns and array of recent bad blocks that the client has seen |
+| debug_getRawBlock        |     block_number      | returns an RLP-encoded block                                    |
+| debug_getRawHeader       |     block_number      | returns an RLP-encoded header                                   |
+| debug_getRawReceipts     |     block_number      | returns an array of EIP-2718 binary-encoded receipts            |
+| debug_getRawTransactions |        tx_hash        | returns an array of EIP-2718 binary-encoded transactions        |
+
+### Engine
```
Comment (taxmeifyoucan) [inline docs/wiki/EL/JSON-RPC.md:41]: I would reorder these in more logical way rather than alphabetically. Start with eth, then debug and finish with engine api, highlight that it's meant for client to client communication, not user client

File: docs/wiki/EL/JSON-RPC.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+  id: 1,
+  headers: {
+    'Content-Type': 'application/json',
+  },
+});
+```
+As you may notice, the JSON-RPC methods are wrapped in a POST request, and the parameters are passed in the body of the request.
+This is a different way to exchange data between the client and the server using the OSI's application layer: the HTTP protocol.
+
+Usually, all the web3 libraries wrap the JSON-RPC methods providing a more friendly way to interact with the execution layer. Please, look forward in your preferred programming language as the syntax may vary.
```
Comment (taxmeifyoucan) [inline docs/wiki/EL/JSON-RPC.md:107]: This is a good point and apart of curl/axios, you should add an example from a library because that's most likely to be used anyway. E.g. web3.py makes it very simple https://web3py.readthedocs.io/en/stable/overview.html#overview
Regarding the axios example, I guess it's nice to have a generic js approach but most of users will go directly with ethers, web3.js or similar libraries

**Comment by rdr-x** [comment]
hey @taxmeifyoucan I just addressed all the comments and suggestions you made. Whenever you have some time take a look at the latest version and tell me what do you think. (:

Considering the main branch was updated a few hours ago I rebase this one so we make sure is up-to-date before merging it.

**Comment by raxhvl** [comment]
This is nice @rdr-x!

**Review by taxmeifyoucan — APPROVED** [review]
Thanks for all the updates! It looks great, I just did minor additions, it's good to go
